nucleus/admin/server-mgmt/src/main/java/fish/payara/admin/servermgmt/cli/PrintCertificateCommand.java

```diff
@@ -0,0 +1,255 @@
+        // PKCS #12: p12, pfx, pkcs12
+        // Other Java Keystores: jks, jceks (both Oracle proprietary)
+        return this.file.trim().matches(".*\\.(jks|jceks|pkcs12|pfx|p12)");
+    }
+
+
+    private String getKeystoreType() {
+        final String ksFilename = this.keystoreFile.getName().toLowerCase();
+        if (ksFilename.endsWith("jks")) {
+            return "JKS";
+        }
+        if (ksFilename.endsWith("jceks")) {
+            return "JCEKS";
+        }
+        if (ksFilename.endsWith("p12") || ksFilename.endsWith("pfx") || ksFilename.endsWith("pkcs12")) {
+            return "PKCS12";
+        }
+        throw new IllegalStateException("Reached unreachable code, validation is incomplete!");
+    }
+
+
+    private X509Certificate getCertificate() throws CommandException {
+        if (this.derFile != null) {
+            return getCertificateFromDerFile();
+        } if (this.keystoreFile != null) {
+            return getCertificateFromKeystore();
+        }
+        throw new CommandException("Could not read the certificate from the provided file.");
+    }
+
+
+    private X509Certificate getCertificateFromDerFile() throws CommandException {
+        try {
+            final CertificateSupplier supplier = () -> {
+                final KeystoreManager manager = new KeystoreManager();
+                final Collection<? extends Certificate> chain = manager.readPemCertificateChain(this.derFile);
+                if (chain.isEmpty()) {
+                    return null;
+                }
+                return chain.iterator().next();
+            };
+            return getX509Certificate(supplier);
+        } catch (final KeyStoreException e) {
+            throw new CommandException("Could not read the certificate from the provided file.", e);
+        }
+    }
+
+
+    private X509Certificate getCertificateFromKeystore() throws CommandException {
+        try {
+            final CertificateSupplier supplier = () -> {
+                final KeystoreManager manager = new KeystoreManager();
+                final String ksType = getKeystoreType();
+                final KeyStore keystore = manager.openKeyStore(this.keystoreFile, ksType, this.keystorePassword);
+                return keystore.getCertificate(this.certificateAlias);
+            };
+            return getX509Certificate(supplier);
+        } catch (final KeyStoreException e) {
+            throw new CommandException("Could not read the certificate from the provided keystore.", e);
+        }
+    }
+
+
+    private static X509Certificate getX509Certificate(final CertificateSupplier supplier) throws KeyStoreException {
+        final Certificate certificate  = supplier.get();
+        if (certificate instanceof X509Certificate) {
+            return (X509Certificate) certificate;
+        }
+        throw new KeyStoreException("The certificate was found but it is not supported X509 certificate.");
+    }
+
+
+    private interface CertificateSupplier {
```

Comment:
You could use `Callable` but it might be better readability you are after.

Reply:
Or - drop this interface definition and use `java.util.function.Supplier`.

Reply:
@pzygielo `java.util.function.Supplier` lacks the checked exception and cannot be used.

Reply:
> @pzygielo `java.util.function.Supplier` lacks the checked exception and cannot be used.

:+1: right, missed that!

Reply:
Supplier does not allow to throw exceptions.

Reply:
Callable throws java.lang.Exception. But yes, because currently I'm not interested in the exception type, there is no need for the interface.